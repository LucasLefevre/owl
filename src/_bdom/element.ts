import { Block } from "./types";

export interface BlockElem extends Block<BlockElem> {
  data: any[];
  children: Block[];
  refs?: (HTMLElement | Text)[];
}

type BlockBuilder = (data?: any[], children?: any[]) => BlockElem;

// const mountBefore = (build: BlockElem["build"]): BlockElem["mountBefore"] =>  {
//   return function mountBefore(this: BlockElem, anchor: any) {
//     this.build();
//     anchor.before(this.el);
//   }
// };

// function  patch(block1: any, block2: any) {
//     if (block1 === block2) {
//       return;
//     }
//     (block1 as any).data.builder.update((block2 as any).data, block2.content);
//   },
function moveBefore(this: BlockElem, anchor: any) {
  anchor.before(this.el!);
}

function remove(this: BlockElem) {
  const el = this.el!;
  el.parentElement!.removeChild(el);
}

function firstChildNode(this: BlockElem): ChildNode | null {
  return this.el;
}

// -----------------------------------------------------------------------------
//  makeBuilder
// -----------------------------------------------------------------------------
interface BuilderContext {
  path: string[];
  isDynamic: boolean;
  hasChild: boolean;
  hasHandler: boolean;
  info: {
    index: number;
    type: "text" | "child" | "handler" | "attribute";
    path: string;
    event?: string;
    name?: string;
  }[];
}

function toDom(node: ChildNode, ctx: BuilderContext): HTMLElement | Text | Comment {
  switch (node.nodeType) {
    case 1: {
      // HTMLElement
      const tagName = (node as Element).tagName;
      if (tagName.startsWith("owl-text-")) {
        const index = parseInt(tagName.slice(9), 10);
        ctx.info.push({ index, path: ctx.path.join("."), type: "text" });
        ctx.isDynamic = true;
        return document.createTextNode("");
      }
      if (tagName.startsWith("owl-child-")) {
        const index = parseInt(tagName.slice(10), 10);
        ctx.info.push({ index, type: "child", path: ctx.path.join(".") });
        ctx.hasChild = true;
        return document.createTextNode("");
      }
      const result = document.createElement((node as Element).tagName);
      const attrs = (node as Element).attributes;
      for (let i = 0; i < attrs.length; i++) {
        const attrName = attrs[i].name;
        const attrValue = attrs[i].value;
        if (attrName.startsWith("owl-handler-")) {
          const index = parseInt(attrName.slice(12), 10);
          ctx.info.push({ index, path: ctx.path.join("."), type: "handler", event: attrValue });
          ctx.hasHandler = true;
        } else if (attrName.startsWith("owl-attribute-")) {
          const index = parseInt(attrName.slice(14), 10);
          ctx.info.push({ index, path: ctx.path.join("."), type: "attribute", name: attrValue });
          ctx.isDynamic = true;
        } else {
          result.setAttribute(attrs[i].name, attrValue);
        }
      }
      let children = (node as Element).childNodes;
      const initialPath = ctx.path.slice();
      let currentPath = initialPath.slice();
      for (let i = 0; i < children.length; i++) {
        currentPath = currentPath.concat(i === 0 ? "firstChild" : "nextSibling");
        ctx.path = currentPath;
        result.appendChild(toDom(children[i], ctx));
      }
      ctx.path = initialPath;
      return result;
    }
    case 3: {
      // text node
      return document.createTextNode(node.textContent!);
    }
    case 8: {
      // comment node
      return document.createComment(node.textContent!);
    }
  }
  throw new Error("boom");
}

interface CompiledOutput {
  template: HTMLElement;
  builder: (
    template: HTMLElement,
    updateClass: any,
    handler: any
  ) => { mountBefore: BlockElem["mountBefore"]; patch: BlockElem["patch"] };
  hasChild: boolean;
}

export function _compileBlock(str: string): CompiledOutput {
  const info: BuilderContext["info"] = [];
  const ctx: BuilderContext = {
    path: ["el"],
    info,
    hasChild: false,
    hasHandler: false,
    isDynamic: false,
  };

  const doc = new DOMParser().parseFromString(str, "text/xml");
  const template = toDom(doc.firstChild!, ctx) as any;
  const isDynamic = ctx.isDynamic;
  const hasChild = ctx.hasChild;
  const hasHandler = ctx.hasHandler;

  // console.log(context)
  // const signature = hasChild ? "data, children" : isDynamic || hasHandler ? "data" : "";
  const code: string[] = [
    `  return {`,
    `    mountBefore(anchor) {`,
    `      this.el = template.cloneNode(true);`,
  ];

  // preparing all internal refs
  if (info.length) {
    for (let i = 0; i < info.length; i++) {
      code.push(`      let ref${i} = this.${info[i].path};`);
    }

    if (isDynamic) {
      code.push(`      const data = this.data;`);
    }
    if (hasChild) {
      code.push(`      const children = this.children;`);
    }
    if (hasHandler) {
    }
    //   code.push(`      const handler = this.handleEvent;`);

    for (let i = 0; i < info.length; i++) {
      const data = info[i];
      const index = info[i].index;
      switch (info[i].type) {
        case "text":
          code.push(`      ref${i}.textContent = data[${info[i].index}];`);
          break;
        case "attribute":
          if (data.name === "class") {
            code.push(`      updateClass(ref${i}, null, data[${index}]);`);
          } else {
            code.push(
              `      let value${index} = data[${index}]; ref${i}.setAttribute("${data.name}", value${index} === true ? "" : value${index});`
            );
          }
          break;
        case "handler":
          code.push(
            `      ref${i}.addEventListener("${info[i].event}", handler.bind(this, ${info[i].index}));`
          );
          break;

        case "child":
          code.push(
            `      let child${index} = children[${index}]; if (child${index}) { child${index}.mountBefore(ref${i}); }`
          );
      }
    }
    code.push(
      `      this.refs = [${info.map((t: any, index: number) => `ref${index}`).join(", ")}];`
    );
  }

  // if (isDynamic || hasHandler) {
  //   code.push(`      this.data = [];`);
  // }
  // if (hasChild) {
  //   code.push(`      this.children = [];`);
  // }
  // if (hasHandler) {
  //   // todo
  //   code.push(`      const handler = this.handleEvent;`);
  //   for (let i = 0; i < info.length; i++) {
  //     if (info[i].type === "handler") {
  //       code.push(
  //         `      ref${i}.addEventListener("${info[i].event}", handler.bind(this, ${info[i].index}));`
  //       );
  //     }
  //   }
  // }

  // end of constructor
  code.push(`      anchor.before(this.el);`);
  // code.push(`    }`);
  code.push(`    },`);
  code.push(`    patch(block) {`);
  if (info.length) {
    code.push(`      if (this === block) return;`);
    if (isDynamic || hasChild) {
      code.push(`      const refs = this.refs;`);
    }
    if (isDynamic || hasHandler) {
      code.push(`      const current = this.data;`);
      code.push(`      const next = block.data;`);
    }
  }
  // update function
  if (info.length) {
    //   if (isDynamic || hasHandler) {
    //   }
    if (hasChild) {
      code.push(`      const children = this.children;`);
      code.push(`      const nextChildren = block.children;`);
    }

    for (let i = 0; i < info.length; i++) {
      const data = info[i];
      const index = data.index;
      switch (data.type) {
        case "text":
          code.push(
            `      if (next[${index}] !== current[${index}]) { refs[${i}].textContent = next[${index}]}`
          );
          break;
        case "attribute":
          if (data.name === "class") {
            code.push(
              `      if (next[${index}] !== current[${index}]) { updateClass(refs[${i}], current[${index}], next[${index}]); }`
            );
          } else {
            code.push(
              `      if (next[${index}] !== current[${index}]) { let value = next[${index}]; refs[${i}].setAttribute("${data.name}", value === true ? "" : value);}`
            );
          }
          break;
        case "child":
          code.push(
            `      let child${index} = children[${index}], nextChild${index} = nextChildren[${index}];`
          );
          code.push(
            `      if (child${index}) { if (nextChild${index}) { child${index}.patch(nextChild${index}); } else { child${index}.remove(); children[${index}] = null; } }`
          );
          code.push(
            `      else if (nextChild${index}) { nextChild${index}.mountBefore(refs[${i}]); children[${index}] = nextChild${index}; }`
          );
      }
    }
  }
  if (isDynamic || hasHandler) {
    code.push(`      this.data = next;`);
  }
  // }
  // code.push(`    }`);
  // if (hasHandler) {
  //   // todo: move handleEvent outside of compiled code (like updateClass)
  //   code.push(`    handleEvent(n) {`);
  //   code.push(`      const handler = this.data[n];`);
  //   code.push(`      if (typeof handler === "function") {`);
  //   code.push(`        handler();`);
  //   code.push(`      } else {`);
  //   code.push(`        const [owner, method] = handler;`);
  //   code.push(`        owner[method]();`);
  //   code.push(`      }`);
  //   code.push(`    }`);
  // }
  code.push(`    }`);
  code.push(`  };`);

  // console.warn(code.join("\n"));
  const wrapper: any = new Function("template, updateClass, handler", code.join("\n"));
  return { template, builder: wrapper, hasChild };
}

export function makeBlock(str: string): BlockBuilder {
  const { template, builder, hasChild } = _compileBlock(str);
  const { mountBefore, patch } = builder(template, updateClass, handler);
  if (hasChild) {
    return (data: any[] = [], children: Block[] = []) => ({
      mountBefore,
      patch,
      moveBefore,
      remove,
      firstChildNode,
      data,
      children,
      el: undefined,
      key: undefined,
    });
  } else {
    return (data: any[] = []) => ({
      mountBefore,
      patch,
      moveBefore,
      remove,
      firstChildNode,
      data,
      children: [],
      el: undefined,
      key: undefined
    });
  }
}

// -----------------------------------------------------------------------------
// helpers
// -----------------------------------------------------------------------------

function updateClass(elem: HTMLElement, prev: any, next: any) {
  prev = prev === undefined ? {} : toClassObj(prev);
  next = next === undefined ? {} : toClassObj(next);
  // remove classes
  for (let c in prev) {
    if (!(c in next)) {
      elem.classList.remove(c);
    }
  }
  // add classes
  for (let c in next) {
    if (!(c in prev)) {
      elem.classList.add(c);
    }
  }
}

function toClassObj(expr: string | number | { [c: string]: any }) {
  const result: { [c: string]: any } = {};

  if (typeof expr === "object") {
    // this is already an object but we may need to split keys:
    // {'a': true, 'b c': true} should become {a: true, b: true, c: true}
    for (let key in expr) {
      const value = expr[key];
      if (value) {
        const words = key.split(/\s+/);
        for (let word of words) {
          result[word] = value;
        }
      }
    }
    return result;
  }
  if (typeof expr !== "string") {
    expr = String(expr);
  }
  // we transform here a list of classes into an object:
  //  'hey you' becomes {hey: true, you: true}
  const str = expr.trim();
  if (!str) {
    return {};
  }
  let words = str.split(/\s+/);
  for (let i = 0, l = words.length; i < l; i++) {
    result[words[i]] = true;
  }
  return result;
}

function handler(this: BlockElem, n: number) {
  const handler = this.data[n];
  if (typeof handler === "function") {
    handler();
  } else {
    const [owner, method] = handler;
    owner[method]();
  }
}
