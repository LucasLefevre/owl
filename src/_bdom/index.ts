import { Block } from "./types";

export { makeBlock } from "./element";
export { list } from "./list";
export { multi } from "./multi";
export { text } from "./text";

// export function beforeRemove(block: Block) {}

// -----------------------------------------------------------------------------
//  BDom main entry points
// -----------------------------------------------------------------------------

export function mount(block: Block, target: HTMLElement) {
  const anchor = document.createTextNode("");
  target.appendChild(anchor);
  block.mountBefore(anchor);
  anchor.remove();
}

// todo: remove this?
export function patch(block1: Block, block2: Block) {
  block1.patch(block2);
}

// todo: remove this?
export function remove(block: Block) {
  // beforeRemove(block);
  block.remove();
}
