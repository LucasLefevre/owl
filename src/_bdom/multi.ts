import { Anchor, Block } from "./types";

export interface BlockMulti extends Block<BlockMulti> {
  anchors: Anchor[] | undefined;
  blocks: (Block | undefined)[];
}

export function multi(blocks: (Block | undefined)[]): BlockMulti {
  return {
    mountBefore,
    patch,
    moveBefore,
    remove,
    firstChildNode,
    anchors: undefined,
    blocks,
  };
}

function mountBefore(this: BlockMulti, anchor: Anchor) {
  const children = this.blocks;
  const anchors = new Array(children.length);
  for (let i = 0, l = children.length; i < l; i++) {
    let child = children[i];
    const childAnchor = document.createTextNode("");
    anchor.before(childAnchor);
    anchors![i] = childAnchor;
    if (child) {
      child.mountBefore(childAnchor);
    }
  }
  this.anchors = anchors;
}

function patch(this: BlockMulti, block: BlockMulti) {
  if (this === block) {
    return;
  }
  const children = this.blocks;
  const newChildren = block.blocks!;
  const anchors = this.anchors!;
  for (let i = 0, l = children.length; i < l; i++) {
    const block = children[i];
    const newBlock = newChildren[i];
    if (block) {
      if (newBlock) {
        block.patch(newBlock);
      } else {
        children[i] = undefined;
        block.remove();
      }
    } else if (newBlock) {
      children[i] = newBlock;
      newBlock.mountBefore(anchors[i]);
    }
  }
}

function moveBefore(this: BlockMulti, anchor: any) {
  const children = this.blocks;
  const anchors = this.anchors!;
  for (let i = 0, l = children.length; i < l; i++) {
    let child = children[i];
    let _anchor = anchors[i];
    anchor.before(_anchor);
    if (child) {
      child.moveBefore(_anchor);
    }
  }
}

function remove(this: BlockMulti) {
  const children = this.blocks;
  const anchors = this.anchors;
  for (let i = 0; i < children.length; i++) {
    const child = children[i];
    if (child) {
      child.remove();
    }
    anchors![i].remove();
  }
}

function firstChildNode(this: BlockMulti): ChildNode | null {
  const children = this.blocks;
  for (let i = 0, l = children.length; i < l; i++) {
    const child = children[i];
    if (child) {
      return child.el || child.firstChildNode();
    }
  }
  return null;
}
