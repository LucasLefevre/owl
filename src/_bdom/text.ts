import { Anchor, Block } from "./types";

export interface BlockText extends Block<BlockText> {
  el: Text | null;
  text: string;
}

export function text(text: string): BlockText {
  return {
    mountBefore,
    patch,
    moveBefore,
    remove,
    firstChildNode,
    el: null,
    text,
  };
}

function mountBefore(this: BlockText, anchor: Anchor) {
  let el = document.createTextNode(this.text);
  this.el = el;
  anchor.before(el);
}

function patch(this: BlockText, block: BlockText) {
  if (this === block) {
    return;
  }
  let text = block.text as any;
  if (this.text !== text) {
    this.text = text;
    this.el!.textContent = text;
  }
}

function moveBefore(this: BlockText, anchor: any) {
  anchor.before(this.el);
}

function remove(this: BlockText) {
  const el = this.el!;
  el.parentElement!.removeChild(el);
}

function firstChildNode(this: BlockText): ChildNode | null {
  return this.el;
}
