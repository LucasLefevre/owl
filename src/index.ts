import { list, mount, multi, patch, remove, text, makeBlock } from "./_bdom";
import { BList } from "./_bdom/list";
import { BMulti } from "./_bdom/multi";
import { BText } from "./_bdom/text";
import { makeBlockClass } from "./_bdom/element";

export const blockDom = {
  list,
  mount,
  multi,
  patch,
  remove,
  text,
  makeBlock,
  BList,
  BMulti,
  BText,
  makeBlockClass
};


// import { App } from "./app";
// import { Component } from "./component";
// import { getCurrent } from "./b_node";

// export { App, Component };

// export async function mount<T extends typeof Component>(
//   C: T,
//   target: HTMLElement
// ): Promise<InstanceType<T>> {
//   const app = new App(C);
//   return app.mount(target);
// }

// export function useComponent(): Component {
//   const current = getCurrent();
//   return current!.component;
// }

// export { status } from "./status";
// export { Portal } from "./misc/portal";
// export { Memo } from "./misc/memo";
// export { xml } from "./tags";
// export { useState } from "./reactivity";
// export { useRef } from "./refs";
// export { EventBus } from "./event_bus";

// export {
//   onWillStart,
//   onMounted,
//   onWillUnmount,
//   onWillUpdateProps,
//   onWillPatch,
//   onPatched,
//   onRender,
//   onDestroyed,
// } from "./lifecycle_hooks";

export const __info__ = {};
