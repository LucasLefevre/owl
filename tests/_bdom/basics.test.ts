import { mount, multi, patch, remove, text, makeBlock as origMakeBlock } from "../../src/_bdom";
import { _compileBlock } from "../../src/_bdom/element";
import { makeTestFixture } from "../helpers";

function makeBlock(str: string) {
  const { builder } = _compileBlock(str);
  expect(builder.toString()).toMatchSnapshot();
  return origMakeBlock(str);
}

//------------------------------------------------------------------------------
// Setup and helpers
//------------------------------------------------------------------------------

let fixture: HTMLElement;

beforeEach(() => {
  fixture = makeTestFixture();
});

afterEach(() => {
  fixture.remove();
});

describe("text and elem blocks", () => {
  test("simple text block", async () => {
    const tree = text("foo");
    expect(tree.el).toBe(null);
    mount(tree, fixture);
    expect(fixture.innerHTML).toBe("foo");
    expect(tree.el).not.toBe(null);
  });

  test("patching a simple text block", async () => {
    const tree = text("foo");
    mount(tree, fixture);
    expect(fixture.innerHTML).toBe("foo");

    patch(tree, text("bar"));
    expect(fixture.innerHTML).toBe("bar");
  });

  test("simple elem block", async () => {
    const block = makeBlock("<div>foo</div>");
    const tree = block();
    expect(tree.el).toBe(undefined);

    mount(tree, fixture);
    expect(tree.el).not.toBe(undefined);
    expect(fixture.innerHTML).toBe("<div>foo</div>");
  });

  test("a text block can be removed", async () => {
    const tree = text("cat");
    expect(fixture.childNodes.length).toBe(0);
    mount(tree, fixture);
    expect(fixture.innerHTML).toBe("cat");
    expect(fixture.childNodes.length).toBe(1);
    remove(tree);
    expect(fixture.innerHTML).toBe("");
    expect(fixture.childNodes.length).toBe(0);
  });

  test("elem block can be removed", async () => {
    const block = makeBlock("<div>foo</div>");
    const tree = block();
    mount(tree, fixture);
    expect(fixture.innerHTML).toBe("<div>foo</div>");
    remove(tree);
    expect(fixture.innerHTML).toBe("");
    expect(fixture.childNodes.length).toBe(0);
  });

  test("simple block with multiple roots", async () => {
    const block1 = makeBlock("<div>foo</div>");
    const block2 = makeBlock("<span>bar</span>");

    const tree = multi([block1(), block2()]);

    mount(tree, fixture);
    expect(fixture.innerHTML).toBe("<div>foo</div><span>bar</span>");
  });

  test("block with dynamic content", async () => {
    const block = makeBlock("<div><p><owl-text-0/></p></div>");
    const tree = block(["foo"]);

    mount(tree, fixture);
    expect(fixture.innerHTML).toBe("<div><p>foo</p></div>");
    patch(tree, block(["bar"]));
    expect(fixture.innerHTML).toBe("<div><p>bar</p></div>");
  });

  test("block with 2 dynamic text nodes", async () => {
    const block = makeBlock("<div><p><owl-text-0/></p><span><owl-text-1/></span></div>");
    const tree = block(["foo", "bar"]);

    mount(tree, fixture);
    expect(fixture.innerHTML).toBe("<div><p>foo</p><span>bar</span></div>");
    patch(tree, block(["appa", "yip yip"]));
    expect(fixture.innerHTML).toBe("<div><p>appa</p><span>yip yip</span></div>");
  });

  test("block with subblock", async () => {
    const block1 = makeBlock("<div><owl-child-0/></div>");
    const block2 = makeBlock("<p>yip yip</p>");
    const tree = block1([], [block2()]);

    mount(tree, fixture);

    expect(fixture.innerHTML).toBe("<div><p>yip yip</p></div>");
  });

  test("block with 2 subblocks", async () => {
    const block1 = makeBlock("<div><owl-child-0/><owl-child-1/></div>");
    const block2 = makeBlock("<p>yip yip</p>");
    const tree = block1([], [block2(), text("appa")]);

    mount(tree, fixture);

    expect(fixture.innerHTML).toBe("<div><p>yip yip</p>appa</div>");
  });

  test("block with subblock with siblings", async () => {
    const block1 = makeBlock("<div><p>1</p><owl-child-0/><p>2</p></div>");
    const block2 = makeBlock("<p>yip yip</p>");

    const tree = block1([], [block2()]);

    mount(tree, fixture);
    expect(fixture.innerHTML).toBe("<div><p>1</p><p>yip yip</p><p>2</p></div>");
  });

  test("block with conditional child", async () => {
    const block1 = makeBlock("<div><p><owl-child-0/></p></div>");
    const block2 = makeBlock("<span>foo</span>");

    const tree = block1();
    mount(tree, fixture);
    expect(fixture.innerHTML).toBe("<div><p></p></div>");

    patch(tree, block1([], [block2()]));
    expect(fixture.innerHTML).toBe("<div><p><span>foo</span></p></div>");

    patch(tree, block1());
    expect(fixture.innerHTML).toBe("<div><p></p></div>");
  });

  test("block with subblock with dynamic content", async () => {
    const block1 = makeBlock("<div><owl-child-0/></div>");
    const block2 = makeBlock("<p><owl-text-0/></p>");
    const tree = block1([], [block2(["yip yip"])]);

    mount(tree, fixture);
    expect(fixture.innerHTML).toBe("<div><p>yip yip</p></div>");

    patch(tree, block1([], [block2(["foo"])]));
    expect(fixture.innerHTML).toBe("<div><p>foo</p></div>");
  });

  test("block with dynamic content and subblock", async () => {
    const block1 = makeBlock("<div><owl-child-0/><p><owl-text-0/></p></div>");
    const block2 = makeBlock("<p>sub block</p>");
    const tree = block1(["yip yip"], [block2()]);

    mount(tree, fixture);
    expect(fixture.innerHTML).toBe("<div><p>sub block</p><p>yip yip</p></div>");

    patch(tree, block1(["foo"], [block2()]));
    expect(fixture.innerHTML).toBe("<div><p>sub block</p><p>foo</p></div>");
  });
});

describe("misc", () => {
  test("reusing a block skips patching process", async () => {
    const foo = text("foo");
    const bar = text("bar");
    let fooCounter = 0;
    let barCounter = 0;
    let fooValue = "foo";
    let barValue = "bar";
    Object.defineProperty(foo, "text", {
      get() {
        fooCounter++;
        return fooValue;
      },
    });
    Object.defineProperty(bar, "text", {
      get() {
        barCounter++;
        return barValue;
      },
      set(val) {
        barValue = val;
      },
    });

    const bdom = multi([foo, bar]);
    mount(bdom, fixture);
    expect(fooCounter).toBe(1);
    expect(barCounter).toBe(1);
    expect(fixture.innerHTML).toBe("foobar");

    patch(bdom, multi([foo, text("otherbar")]));
    expect(fixture.innerHTML).toBe("foootherbar");
    expect(fooCounter).toBe(1);
    expect(barCounter).toBe(2);
  });
});
