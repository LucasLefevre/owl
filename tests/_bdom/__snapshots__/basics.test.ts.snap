// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`text and elem blocks block with 2 dynamic text nodes 1`] = `
"function anonymous(BaseBlock, template, updateClass, handler
) {
  return class BElem extends BaseBlock {
    mountBefore(anchor) {
      this.el = template.cloneNode(true);
      let ref0 = this.el.firstChild.firstChild;
      let ref1 = this.el.firstChild.nextSibling.firstChild;
      const data = this.data;
      ref0.textContent = data[0];
      ref1.textContent = data[1];
      this.refs = [ref0, ref1];
      anchor.before(this.el);
    }
    patch(block) {
      if (this === block) return;
      const refs = this.refs;
      const current = this.data;
      const next = block.data;
      if (next[0] !== current[0]) { refs[0].textContent = next[0]}
      if (next[1] !== current[1]) { refs[1].textContent = next[1]}
      this.data = next;
    }
  };
}"
`;

exports[`text and elem blocks block with 2 subblocks 1`] = `
"function anonymous(BaseBlock, template, updateClass, handler
) {
  return class BElem extends BaseBlock {
    mountBefore(anchor) {
      this.el = template.cloneNode(true);
      let ref0 = this.el.firstChild;
      let ref1 = this.el.firstChild.nextSibling;
      const children = this.children;
      let child0 = children[0]; if (child0) { child0.mountBefore(ref0); }
      let child1 = children[1]; if (child1) { child1.mountBefore(ref1); }
      this.refs = [ref0, ref1];
      anchor.before(this.el);
    }
    patch(block) {
      if (this === block) return;
      const refs = this.refs;
      const children = this.children;
      const nextChildren = block.children;
      let child0 = children[0], nextChild0 = nextChildren[0];
      if (child0) { if (nextChild0) { child0.patch(nextChild0); } else { child0.remove(); children[0] = null; } }
      else if (nextChild0) { nextChild0.mountBefore(refs[0]); children[0] = nextChild0; }
      let child1 = children[1], nextChild1 = nextChildren[1];
      if (child1) { if (nextChild1) { child1.patch(nextChild1); } else { child1.remove(); children[1] = null; } }
      else if (nextChild1) { nextChild1.mountBefore(refs[1]); children[1] = nextChild1; }
    }
  };
}"
`;

exports[`text and elem blocks block with 2 subblocks 2`] = `
"function anonymous(BaseBlock, template, updateClass, handler
) {
  return class BElem extends BaseBlock {
    mountBefore(anchor) {
      this.el = template.cloneNode(true);
      anchor.before(this.el);
    }
    patch(block) {
    }
  };
}"
`;

exports[`text and elem blocks block with conditional child 1`] = `
"function anonymous(BaseBlock, template, updateClass, handler
) {
  return class BElem extends BaseBlock {
    mountBefore(anchor) {
      this.el = template.cloneNode(true);
      let ref0 = this.el.firstChild.firstChild;
      const children = this.children;
      let child0 = children[0]; if (child0) { child0.mountBefore(ref0); }
      this.refs = [ref0];
      anchor.before(this.el);
    }
    patch(block) {
      if (this === block) return;
      const refs = this.refs;
      const children = this.children;
      const nextChildren = block.children;
      let child0 = children[0], nextChild0 = nextChildren[0];
      if (child0) { if (nextChild0) { child0.patch(nextChild0); } else { child0.remove(); children[0] = null; } }
      else if (nextChild0) { nextChild0.mountBefore(refs[0]); children[0] = nextChild0; }
    }
  };
}"
`;

exports[`text and elem blocks block with conditional child 2`] = `
"function anonymous(BaseBlock, template, updateClass, handler
) {
  return class BElem extends BaseBlock {
    mountBefore(anchor) {
      this.el = template.cloneNode(true);
      anchor.before(this.el);
    }
    patch(block) {
    }
  };
}"
`;

exports[`text and elem blocks block with dynamic content 1`] = `
"function anonymous(BaseBlock, template, updateClass, handler
) {
  return class BElem extends BaseBlock {
    mountBefore(anchor) {
      this.el = template.cloneNode(true);
      let ref0 = this.el.firstChild.firstChild;
      const data = this.data;
      ref0.textContent = data[0];
      this.refs = [ref0];
      anchor.before(this.el);
    }
    patch(block) {
      if (this === block) return;
      const refs = this.refs;
      const current = this.data;
      const next = block.data;
      if (next[0] !== current[0]) { refs[0].textContent = next[0]}
      this.data = next;
    }
  };
}"
`;

exports[`text and elem blocks block with dynamic content and subblock 1`] = `
"function anonymous(BaseBlock, template, updateClass, handler
) {
  return class BElem extends BaseBlock {
    mountBefore(anchor) {
      this.el = template.cloneNode(true);
      let ref0 = this.el.firstChild;
      let ref1 = this.el.firstChild.nextSibling.firstChild;
      const data = this.data;
      const children = this.children;
      let child0 = children[0]; if (child0) { child0.mountBefore(ref0); }
      ref1.textContent = data[0];
      this.refs = [ref0, ref1];
      anchor.before(this.el);
    }
    patch(block) {
      if (this === block) return;
      const refs = this.refs;
      const current = this.data;
      const next = block.data;
      const children = this.children;
      const nextChildren = block.children;
      let child0 = children[0], nextChild0 = nextChildren[0];
      if (child0) { if (nextChild0) { child0.patch(nextChild0); } else { child0.remove(); children[0] = null; } }
      else if (nextChild0) { nextChild0.mountBefore(refs[0]); children[0] = nextChild0; }
      if (next[0] !== current[0]) { refs[1].textContent = next[0]}
      this.data = next;
    }
  };
}"
`;

exports[`text and elem blocks block with dynamic content and subblock 2`] = `
"function anonymous(BaseBlock, template, updateClass, handler
) {
  return class BElem extends BaseBlock {
    mountBefore(anchor) {
      this.el = template.cloneNode(true);
      anchor.before(this.el);
    }
    patch(block) {
    }
  };
}"
`;

exports[`text and elem blocks block with subblock 1`] = `
"function anonymous(BaseBlock, template, updateClass, handler
) {
  return class BElem extends BaseBlock {
    mountBefore(anchor) {
      this.el = template.cloneNode(true);
      let ref0 = this.el.firstChild;
      const children = this.children;
      let child0 = children[0]; if (child0) { child0.mountBefore(ref0); }
      this.refs = [ref0];
      anchor.before(this.el);
    }
    patch(block) {
      if (this === block) return;
      const refs = this.refs;
      const children = this.children;
      const nextChildren = block.children;
      let child0 = children[0], nextChild0 = nextChildren[0];
      if (child0) { if (nextChild0) { child0.patch(nextChild0); } else { child0.remove(); children[0] = null; } }
      else if (nextChild0) { nextChild0.mountBefore(refs[0]); children[0] = nextChild0; }
    }
  };
}"
`;

exports[`text and elem blocks block with subblock 2`] = `
"function anonymous(BaseBlock, template, updateClass, handler
) {
  return class BElem extends BaseBlock {
    mountBefore(anchor) {
      this.el = template.cloneNode(true);
      anchor.before(this.el);
    }
    patch(block) {
    }
  };
}"
`;

exports[`text and elem blocks block with subblock with dynamic content 1`] = `
"function anonymous(BaseBlock, template, updateClass, handler
) {
  return class BElem extends BaseBlock {
    mountBefore(anchor) {
      this.el = template.cloneNode(true);
      let ref0 = this.el.firstChild;
      const children = this.children;
      let child0 = children[0]; if (child0) { child0.mountBefore(ref0); }
      this.refs = [ref0];
      anchor.before(this.el);
    }
    patch(block) {
      if (this === block) return;
      const refs = this.refs;
      const children = this.children;
      const nextChildren = block.children;
      let child0 = children[0], nextChild0 = nextChildren[0];
      if (child0) { if (nextChild0) { child0.patch(nextChild0); } else { child0.remove(); children[0] = null; } }
      else if (nextChild0) { nextChild0.mountBefore(refs[0]); children[0] = nextChild0; }
    }
  };
}"
`;

exports[`text and elem blocks block with subblock with dynamic content 2`] = `
"function anonymous(BaseBlock, template, updateClass, handler
) {
  return class BElem extends BaseBlock {
    mountBefore(anchor) {
      this.el = template.cloneNode(true);
      let ref0 = this.el.firstChild;
      const data = this.data;
      ref0.textContent = data[0];
      this.refs = [ref0];
      anchor.before(this.el);
    }
    patch(block) {
      if (this === block) return;
      const refs = this.refs;
      const current = this.data;
      const next = block.data;
      if (next[0] !== current[0]) { refs[0].textContent = next[0]}
      this.data = next;
    }
  };
}"
`;

exports[`text and elem blocks block with subblock with siblings 1`] = `
"function anonymous(BaseBlock, template, updateClass, handler
) {
  return class BElem extends BaseBlock {
    mountBefore(anchor) {
      this.el = template.cloneNode(true);
      let ref0 = this.el.firstChild.nextSibling;
      const children = this.children;
      let child0 = children[0]; if (child0) { child0.mountBefore(ref0); }
      this.refs = [ref0];
      anchor.before(this.el);
    }
    patch(block) {
      if (this === block) return;
      const refs = this.refs;
      const children = this.children;
      const nextChildren = block.children;
      let child0 = children[0], nextChild0 = nextChildren[0];
      if (child0) { if (nextChild0) { child0.patch(nextChild0); } else { child0.remove(); children[0] = null; } }
      else if (nextChild0) { nextChild0.mountBefore(refs[0]); children[0] = nextChild0; }
    }
  };
}"
`;

exports[`text and elem blocks block with subblock with siblings 2`] = `
"function anonymous(BaseBlock, template, updateClass, handler
) {
  return class BElem extends BaseBlock {
    mountBefore(anchor) {
      this.el = template.cloneNode(true);
      anchor.before(this.el);
    }
    patch(block) {
    }
  };
}"
`;

exports[`text and elem blocks elem block can be removed 1`] = `
"function anonymous(BaseBlock, template, updateClass, handler
) {
  return class BElem extends BaseBlock {
    mountBefore(anchor) {
      this.el = template.cloneNode(true);
      anchor.before(this.el);
    }
    patch(block) {
    }
  };
}"
`;

exports[`text and elem blocks simple block with multiple roots 1`] = `
"function anonymous(BaseBlock, template, updateClass, handler
) {
  return class BElem extends BaseBlock {
    mountBefore(anchor) {
      this.el = template.cloneNode(true);
      anchor.before(this.el);
    }
    patch(block) {
    }
  };
}"
`;

exports[`text and elem blocks simple block with multiple roots 2`] = `
"function anonymous(BaseBlock, template, updateClass, handler
) {
  return class BElem extends BaseBlock {
    mountBefore(anchor) {
      this.el = template.cloneNode(true);
      anchor.before(this.el);
    }
    patch(block) {
    }
  };
}"
`;

exports[`text and elem blocks simple elem block 1`] = `
"function anonymous(BaseBlock, template, updateClass, handler
) {
  return class BElem extends BaseBlock {
    mountBefore(anchor) {
      this.el = template.cloneNode(true);
      anchor.before(this.el);
    }
    patch(block) {
    }
  };
}"
`;
